src/RootHitFile.jl

```diff
@@ -0,0 +1,86 @@
+on energy depositions, i.e. hits.
+
+A `RootHitFile` is also a Tables.jl compatible row table of events. `batch_size` determines
+the number of events grouped into partitions when `Tables.partitions` is used. To change the
+default of 10, see `DEFAULTS`.
+"""
+struct RootHitFile
+    stream::IOBuffer
+    batch_size::Int
+
+    function RootHitFile(stream::IOStream; batch_size=DEFAULTS.root_hits_batch_size)
+        new(IOBuffer(mmap(stream)), batch_size)
+    end
+end
+
+function RootHitFile(path::AbstractString; batch_size=DEFAULTS.root_hits_batch_size)
+    if occursin(r".root.hits$", path)
+        return RootHitFile(open(path); batch_size=batch_size)
+    else
+        throw(ArgumentError("$path is not a .root.hits file"))
+    end
+end
+
+EventTuple = NamedTuple{
+    (:eventnum, :primcount, :pos, :E, :time, :particleID, :trkID, :trkparentID, :volumeID),
+    Tuple{
+        Int32, Int32, Vector{NTuple{3, Float32}}, Vector{Float32}, Vector{Float32},
+        Vector{Int32}, Vector{Int32}, Vector{Int32}, Vector{String}
+    }
+}
+
+function Base.read(f::RootHitFile)
+    eventnum  = Parsers.parse(Int32, f.stream)
+    hitcount  = Parsers.parse(Int32, f.stream)
+    primcount = Parsers.parse(Int32, f.stream)
+
+    # skip newline
+    skip(f.stream, 1)
+
+    pos         = Vector{NTuple{3, Float32}}(undef, hitcount)
+    E           = Vector{            Float32}(undef, hitcount)
+    time        = Vector{            Float32}(undef, hitcount)
+    particleID  = Vector{              Int32}(undef, hitcount)
+    trkID       = Vector{              Int32}(undef, hitcount)
+    trkparentID = Vector{              Int32}(undef, hitcount)
+    volumeID    = Vector{             String}(undef, hitcount)
+
+    @inbounds for i in 1:hitcount
+        pos[i] = (
+            Parsers.parse(Float32, f.stream),
+            Parsers.parse(Float32, f.stream),
+            Parsers.parse(Float32, f.stream)
+        )
+        E[i]           = Parsers.parse(Float32, f.stream)
+        time[i]        = Parsers.parse(Float32, f.stream)
+        particleID[i]  = Parsers.parse(Int32, f.stream)
+        trkID[i]       = Parsers.parse(Int32, f.stream)
+        trkparentID[i] = Parsers.parse(Int32, f.stream)
+        volumeID[i]    = String(readuntil(f.stream, UInt8('\n')))
+    end
+
+    return EventTuple((
+        eventnum, primcount, pos, E, time,
+        particleID, trkID, trkparentID, volumeID
+    ))
+end
+
+Base.eof(f::RootHitFile) = eof(f.stream)
+
+function Base.iterate(f::RootHitFile, state = nothing)
+    eof(f) && return nothing
+    return read(f), nothing
+end
+
+Base.IteratorSize(::Type{RootHitFile}) = Base.SizeUnknown()
+Base.IteratorEltype(::Type{RootHitFile}) = Base.HasEltype()
+Base.eltype(::Type{RootHitFile}) = EventTuple
+
+Tables.isrowtable(::Type{RootHitFile}) = true
+Tables.schema(f::RootHitFile) = Tables.Schema(eltype(f))
+
+Tables.partitions(f::RootHitFile) = Iterators.partition(f, f.batch_size)
```

Comment:
Please to return an iterator over `TypedTables.Table` or `StructArrays.StructArray` here, depending on whether the (renamed) `EventTuple` will be a `NamedTuple` or `struct`.

Reply:
Ok, I started working on this a bit. This would require quite a bit of additional code for a feature that the main user of this piece of code (me) probably won't use. I'd suggest leaving this out of this PR for now. Of course, if people start using this and want this feature, I'd be happy to help implement this or give guidance, but right now I don't see this having adequate return on investment.

Let me know if you'd prefer to keep the partition operator the way it's now or if you want to just get rid of it and leave it open to implement later.

Reply:
Why would it require so much extra code to include the conversion to a `TypedTables.Table` in the partitions iterator? Since we're going with `NamedTuple`s, we won't use `StructArray`s here.

Reply:
As I see it I'd need / want to change three aspects of the code, for the purpose of making efficient column based partitioned iteration work.

1. To make this efficient I can't use the `Iterators.partition` because it is an eager iterator and returns vectors of `eltype` as partitions. I want to write directly into a TypedTable, rather than a vector of named tuples (and then converting to TypedTable).
2. In order to not return a vector of named tuples (and convert to TypedTable after), I need to write my own partition iterator class (something like `PartionedRootHitFile`) where I define how it should allocate an empty table and then write into it on each iteration of the underlying file until the partition is full or the file has ended.
3. To make writing into a VectorOfVector work efficiently I have to add a mutating `read!` function which will read into some kind of NamedTuple. Then I can use the EventTuple I defined in the normal case for `read` and a special tuple, using VectorOfVectors for the partitioned case. If I didn't do this it would write into a `Vector` and then have to copy the vector into the relevant part of the `VectorOfVector`.

Assuming I didn't miss some obvious way to do this in a much simpler way, I'd say that these changes should be moved to a separate PR someday. All of these are reasonable additions to make, but it's a significant bit of code to write and test. If you like we can open an issue for it now and lay out this plan so that if, in future, people come across this and want the feature they can see that this was discussed.

Note, if you want I can add a mutating `read!`, since that's a nice thing to have regardless.

Reply:
Well, this actually doesn't have to be super performant, because the parsing will still be the step that limits speed. But sure, we can do this later on.